### PR TITLE
mon: enable different backends (e.g., rocksdb)

### PR DESCRIPTION
**Pull request by liewegas**
- persist the backend type
- fix rocksdb options

https://github.com/ceph/ceph-qa-suite/pull/979

### CODE REVIEW

File: src/mon/MonitorDBStore.h [hunk 1]
```diff
@@ -577,11 +579,58 @@ class MonitorDBStore
...
+      assert(0 != "MonitorDBStore: error initializing keyvaluedb back storage");
+    }
+    db.reset(db_ptr);
+
+    if (g_conf->mon_debug_dump_transactions) {
+      if (!g_conf->mon_debug_dump_json) {
+        dump_fd_binary = ::open(
+          g_conf->mon_debug_dump_location.c_str(),
+          O_CREAT|O_APPEND|O_WRONLY, 0644);
+        if (!dump_fd_binary) {
```
Comment (xiexingguo) [inline src/mon/MonitorDBStore.h]: Should be: 
```c++
if (dump_fd_binary < 0)

**Comment by badone** [comment]
@liewegas Can we explain the test failures?

**Comment by tchaikov** [comment]
there was something leftover in mkfs/a after removing mkfs/a/store.db. not sure what or why...

**Comment by tchaikov** [comment]
we should have
```diff
diff --git a/src/test/mon/mkfs.sh b/src/test/mon/mkfs.sh
index 324457a..a754873 100755
--- a/src/test/mon/mkfs.sh
+++ b/src/test/mon/mkfs.sh
@@ -132,6 +132,7 @@ function auth_cephx_key() {
         return 1
     else
         rm -fr $MON_DIR/store.db
+        rm -f $MON_DIR/kv_backend
     fi

     mon_mkfs --key=$key
```
also to cleanup the kv_backend leftover.

File: src/mon/MonitorDBStore.h [hunk 2]
```diff
@@ -635,8 +640,16 @@ class MonitorDBStore
...
 
   int create_and_open(ostream &out) {
+    // record the type before open
+    string kv_type;
+    int r = read_meta("kv_backend", &kv_type);
+    if (r < 0) {
+      r = write_meta("kv_backend", g_conf->mon_keyvaluedb);
+      if (r < 0)
+	return r;
+    }
```
Comment (tchaikov) [inline src/mon/MonitorDBStore.h]: can we pass the `kv_type` to `_open()`, so we don't need to read it again. and it helps to consolidate the logic handling missing or empty kv_type.

File: src/mon/MonitorDBStore.h [hunk 3]
```diff
@@ -621,51 +675,65 @@ class MonitorDBStore
...
+   *
+   * Read a simple key value to an unopened/mounted store.
+   *
+   * Trailing whitespace is stripped off.
+   *
+   * @param key key name
+   * @param value pointer to value string
+   * @returns 0 for success, or an error code
+   */
+  int read_meta(const std::string& key,
```
Comment (tchaikov) [inline src/mon/MonitorDBStore.h]: could make this method a `const` one.

File: src/mon/MonitorDBStore.h [hunk 4]
```diff
@@ -621,51 +675,65 @@ class MonitorDBStore
...
+   * The default implementation writes this to a plaintext file in the
+   * path.
+   *
+   * A newline is appended.
+   *
+   * @param key key name (e.g., "fsid")
+   * @param value value (e.g., a uuid rendered as a string)
+   * @returns 0 for success, or an error code
+   */
+  int write_meta(const std::string& key,
```
Comment (tchaikov) [inline src/mon/MonitorDBStore.h]: could make this method a `const` one.

**Comment by liewegas** [comment]
Updated.  Also moved the dump fd bug fix to the beginning so we can backport it.

File: src/mon/MonitorDBStore.h [hunk 5]
```diff
@@ -635,8 +640,16 @@ class MonitorDBStore
   }
 
   int create_and_open(ostream &out) {
-    _open();
-    int r = db->create_and_open(out);
+    // record the type before open
+    string kv_type;
+    int r = read_meta("kv_backend", &kv_type);
+    if (r < 0) {
+      r = write_meta("kv_backend", g_conf->mon_keyvaluedb);
```
Comment (tchaikov) [inline src/mon/MonitorDBStore.h]: might want to assign `g_conf->mon_keyvaluedb` to `kv_type` here?

**Comment by tchaikov** [comment]
lgtm

**Comment by tchaikov** [comment]
the test failure is addressed by https://github.com/ceph/ceph/pull/8907

**Comment by badone** [comment]
Looks fine to me

**Pull request merged**